flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/SortBufferAccumulator.java

```diff
@@ -94,29 +99,58 @@ public class SortBufferAccumulator implements BufferAccumulator {
     @Nullable
     private TriConsumer<TieredStorageSubpartitionId, Buffer, Integer> accumulatedBufferFlusher;
 
+    /**
+     * An executor to periodically check the size of buffer pool. If the size is changed, the
+     * accumulated buffers should be flushed to release the buffers.
+     */
+    private final ScheduledExecutorService periodicalAccumulatorFlusher =
+            Executors.newSingleThreadScheduledExecutor(
+                    new ExecutorThreadFactory("hybrid-shuffle-periodical-accumulator-flusher"));
+
+    private final long poolSizeCheckInterval;
+
+    private AtomicInteger poolSize;
+
     /** Whether the current {@link DataBuffer} is a broadcast sort buffer. */
     private boolean isBroadcastDataBuffer;
 
     public SortBufferAccumulator(
             int numSubpartitions,
-            int numBuffers,
+            int numExpectedBuffers,
             int bufferSizeBytes,
+            long poolSizeCheckInterval,
             TieredStorageMemoryManager memoryManager,
             boolean isPartialRecordAllowed) {
         this.numSubpartitions = numSubpartitions;
         this.bufferSizeBytes = bufferSizeBytes;
-        this.numBuffers = numBuffers;
+        this.numExpectedBuffers = numExpectedBuffers;
+        this.poolSizeCheckInterval = poolSizeCheckInterval;
         this.memoryManager = memoryManager;
         this.isPartialRecordAllowed = isPartialRecordAllowed;
     }
 
     @Override
     public void setup(TriConsumer<TieredStorageSubpartitionId, Buffer, Integer> bufferFlusher) {
         this.accumulatedBufferFlusher = bufferFlusher;
+        this.poolSize = new AtomicInteger(memoryManager.getBufferPoolSize());
+
+        if (poolSizeCheckInterval > 0) {
+            periodicalAccumulatorFlusher.scheduleAtFixedRate(
+                    () -> {
+                        int newSize = this.memoryManager.getBufferPoolSize();
+                        int oldSize = poolSize.getAndSet(newSize);
+                        if (oldSize != newSize) {
```

Comment:
Because we need to make sure some buffers can be recycled to meed the requirement of buffers pools whose size is increased. However, I will change the if statement to `if (oldSize > newSize)`.